### ISSUE TITLE
Build of extra-utils (in stage-1) fails

### ISSUE THREAD

**Issue by mhwombat** [NixOS/nixpkgs#203780]
### Steps To Reproduce
1. build *X*

### Build log

Full log available at https://gist.github.com/mhwombat/d2ec165227b1d0a6adb20f29dbb3b6c3

```
patching /nix/store/g13rzqqimgclvzb1414l8877k4qa7c9d-extra-utils/lib/libext2fs.so.2...
patchelf: missing ELF header
error: builder for '/nix/store/8ryi6wva8sycx228dzdg938ir50ivmp6-extra-utils.drv' failed with exit code 1
error: 1 dependencies of derivation '/nix/store/xyfzs6b6hv4gpgc1glpw69pdrch6pgiy-stage-1-init.sh.drv' failed to build
error: 1 dependencies of derivation '/nix/store/s8pg4z5a0s4hr7aq83vj15dipi69n7pl-initrd-linux-5.15.79.drv' failed to build
error: 1 dependencies of derivation '/nix/store/anv9387fbpy95pkpjkw7idsgrjvnd69h-nixos-system-wombat11k-23.05pre429868.a115bb9bd56.drv' failed to build
```

### Notify maintainers

@pennae
@klemensn 
@trofi 

### Metadata

```console
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.76, NixOS, 22.11 (Raccoon), 22.11pre423448.a2a777538d9`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by trofi** [comment]
> patching /nix/store/g13rzqqimgclvzb1414l8877k4qa7c9d-extra-utils/lib/libext2fs.so.2...
> patchelf: missing ELF header

That looks odd. Can you extract that file and check if it's an `ELF` by running `file` on it? `--keep-failed` is probably the simplest way to get to it.

**Comment by mhwombat** [comment]
Hmmm... the file is empty!

```
nixos-config on  main [!?] via ❄️  impure (shell) 
❯ file /nix/store/g13rzqqimgclvzb1414l8877k4qa7c9d-extra-utils/lib/libext2fs.so.2
/nix/store/g13rzqqimgclvzb1414l8877k4qa7c9d-extra-utils/lib/libext2fs.so.2: empty

nixos-config on  main [!?] via ❄️  impure (shell) 
❯ ls -l /nix/store/g13rzqqimgclvzb1414l8877k4qa7c9d-extra-utils/lib/libext2fs.so.2
-rwxr-xr-x 1 nixbld10 nixbld 0 Nov 30 19:23 /nix/store/g13rzqqimgclvzb1414l8877k4qa7c9d-extra-utils/lib/libext2fs.so.2
```

**Comment by trofi** [comment]
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/stage-1.nix#L93-L261 should be the full definition of `extra-utils`.

I guess it happens for setups that use `ext*` filesystem:

```nix
      # Copy resize2fs if any ext* filesystems are to be resized
      ${optionalString (any (fs: fs.autoResize && (lib.hasPrefix "ext" fs.fsType)) fileSystems) ''
        # We need mke2fs in the initrd.
        copy_bin_and_libs ${pkgs.e2fsprogs}/sbin/resize2fs
      ''}
```

I didn't have any `ext` filesystems locally. When I tweak it to enable locally unconditionally I still don't get the failure:

```diff
--- a/nixos/modules/system/boot/stage-1.nix
+++ b/nixos/modules/system/boot/stage-1.nix
@@ -171,7 +171,7 @@ let
       chmod +x $out/bin/modprobe-kernel

       # Copy resize2fs if any ext* filesystems are to be resized
-      ${optionalString (any (fs: fs.autoResize && (lib.hasPrefix "ext" fs.fsType)) fileSystems) ''
+      ${optionalString true ''
         # We need mke2fs in the initrd.
         copy_bin_and_libs ${pkgs.e2fsprogs}/sbin/resize2fs
       ''}
```

Try to add a few debug statements there to check how it decides to copy `libext2fs.so.2` and if the source is an empty file (it should not be). I suspect you have something else on top to trigger that script to do something unusual.

**Comment by mhwombat** [comment]
**TL;DR: I spotted two things that I think might be problems:**
1. The condition evaluates to false, so `copy_bin_and_libs` isn't executed, and
2. Even if I fix that, the file that `libext2fs.so.2` is supposed to be a copy of is empty.

---

I added some debug like so:

```
diff --git a/nixos/modules/system/boot/stage-1.nix b/nixos/modules/system/boot/stage-1.nix
index 28c76fb169f..04bcd7c6e2f 100644
--- a/nixos/modules/system/boot/stage-1.nix
+++ b/nixos/modules/system/boot/stage-1.nix
@@ -170,11 +170,14 @@ let
       EOF
       chmod +x $out/bin/modprobe-kernel
 
+      echo "DEBUG 1"
       # Copy resize2fs if any ext* filesystems are to be resized
       ${optionalString (any (fs: fs.autoResize && (lib.hasPrefix "ext" fs.fsType)) fileSystems) ''
+        echo "DEBUG 2"
         # We need mke2fs in the initrd.
         copy_bin_and_libs ${pkgs.e2fsprogs}/sbin/resize2fs
       ''}
+      echo "DEBUG 3"
 
       # Copy multipath.
       ${optionalString config.services.multipath.enable ''
```

With this change, in the log I see "DEBUG 1" immediately followed by "DEBUG 3", so **`copy_bin_and_libs` isn't executed** and I get the same error about 

Assuming it's supposed to be executed, I made it unconditional like so:

```
diff --git a/nixos/modules/system/boot/stage-1.nix b/nixos/modules/system/boot/stage-1.nix
index 28c76fb169f..a7ae8673b1b 100644
--- a/nixos/modules/system/boot/stage-1.nix
+++ b/nixos/modules/system/boot/stage-1.nix
@@ -170,11 +170,14 @@ let
       EOF
       chmod +x $out/bin/modprobe-kernel
 
+      echo "DEBUG 1"
       # Copy resize2fs if any ext* filesystems are to be resized
-      ${optionalString (any (fs: fs.autoResize && (lib.hasPrefix "ext" fs.fsType)) fileSystems) ''
+      ${optionalString true ''
+        echo "DEBUG 2"
         # We need mke2fs in the initrd.
         copy_bin_and_libs ${pkgs.e2fsprogs}/sbin/resize2fs
       ''}
+      echo "DEBUG 3"
 
       # Copy multipath.
       ${optionalString config.services.multipath.enable ''
```

Whereupon I got this:

```
DEBUG 1
DEBUG 2
'/nix/store/y3vci6gggcl2z46dkgajvixxpfvvybky-e2fsprogs-1.46.5-bin/sbin/resize2fs' -> '/nix/store/0m7ip6w7n8ajr5ixl5i93d1kw1dv3w6r-extra-utils/bin/resize2fs'
DEBUG 3
<snip>
'/nix/store/47hjj5alk43aws4z7zpps1ngi9fggh0j-e2fsprogs-1.46.5/lib/libext2fs.so.2.4' -> '/nix/store/0m7ip6w7n8ajr5ixl5i93d1kw1dv3w6r-extra-utils/lib/libext2fs.so.2'
<snip>
patching /nix/store/0m7ip6w7n8ajr5ixl5i93d1kw1dv3w6r-extra-utils/lib/libext2fs.so.2...
patchelf: missing ELF header
error: builder for '/nix/store/nfmc29l8m3i5milq7fxfj1w0fqay8iyb-extra-utils.drv' failed with exit code 1
error: 1 dependencies of derivation '/nix/store/q5kknxc1s4lnaj7fcym2pbi76dxqh28x-stage-1-init.sh.drv' failed to build
error: 1 dependencies of derivation '/nix/store/nvincmq47m5lvlg4a6d0spn46rkpjydj-initrd-linux-5.15.80.drv' failed to build
error (ignored): error: cannot unlink '/tmp/nix-build-hplip-3.22.6.drv-15/hplip-3.22.6': Directory not empty
error: 1 dependencies of derivation '/nix/store/5h02ncqkw3mgpnxcs3cm5mdrqwsk7i53-nixos-system-wombat11k-23.05.git.c79f09b7e5bM.drv' failed to build
```

So if I understand this correctly, `libext2fs.so.2` is supposed to be a copy of the file below, which is empty:

```
❯ ls -l /nix/store/47hjj5alk43aws4z7zpps1ngi9fggh0j-e2fsprogs-1.46.5/lib/libext2fs.so.2.4
-r-xr-xr-x 1 root root 0 Jan  1  1970 /nix/store/47hjj5alk43aws4z7zpps1ngi9fggh0j-e2fsprogs-1.46.5/lib/libext2fs.so.2.4
```

**Comment by trofi** [comment]
Aha, good debugging progress! If I build the same derivation I get non-empty file:

```
$ nix build /nix/store/47hjj5alk43aws4z7zpps1ngi9fggh0j-e2fsprogs-1.46.5
$ LANG=C ls -lh /nix/store/47hjj5alk43aws4z7zpps1ngi9fggh0j-e2fsprogs-1.46.5/lib/libext2fs.so.2.4
-r-xr-xr-x 2 root root 476K Jan  1  1970 /nix/store/47hjj5alk43aws4z7zpps1ngi9fggh0j-e2fsprogs-1.46.5/lib/libext2fs.so.2.4
```

That probably means your derivation is built incorrectly or got corrupted on disk (sometimes filesystems zero-out files on disk on abrupt shutdowns before files land on disk). Does `nix` complain about your store integrity?

Simplest way to check it is:

```
$ nix store verify /nix/store/47hjj5alk43aws4z7zpps1ngi9fggh0j-e2fsprogs-1.46.5
$ nix store verify --all
```

Another thing you can try is to rebuild the package to see the difference:

```
$ nix build github:NixOS/nixpkgs/a2a777538d971c6b01c6e54af89ddd6567c055e8#e2fsprogs
$ nix build github:NixOS/nixpkgs/a2a777538d971c6b01c6e54af89ddd6567c055e8#e2fsprogs --rebuild
$ LANG=C ls -lh result/lib/libext2fs.so.2.4
-r-xr-xr-x 2 root root 476K Jan  1  1970 result/lib/libext2fs.so.2.4
```

If you get something else back we can dive into more details.

**Comment by mhwombat** [comment]
Woohoo! That was the problem. Fixed with `nix store repair`. Build is successful now.
